pdns/pdns_recursor.cc

```diff
@@ -1188,9 +1188,16 @@ static void startDoResolve(void *p)
          OPT record.  This MUST also occur when a truncated response (using
          the DNS header's TC bit) is returned."
       */
-      if (addRecordToPacket(pw, makeOpt(edo.d_packetsize, 0, edo.d_Z), minTTL, maxanswersize)) {
-        pw.commit();
+      DNSPacketWriter::optvect_t opts;
+      if(wantsNSID) {
+        const static string mode_server_id = ::arg()["server-id"];
+        if(mode_server_id != "disabled" && !mode_server_id.empty()) {
+          opts.push_back(make_pair(3, mode_server_id));
+          variableAnswer = true; // Can't packetcache an answer with NSID
+        }
       }
+      pw.addOpt(maxanswersize, 0, DNSSECOK ? EDNSOpts::DNSSECOK : 0, opts);
```

Comment:
I don't think we want `maxanswersize` here, since it's capped by the value advertised by the requestor. However the current code seems to just echo back the value advertised by the requestor (`edo.d_packetsize`) which is just wrong. I guess we should use `g_udpTruncationThreshold` instead?